hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/balancer/TestContainerBalancerDatanodeNodeLimit.java

```diff
@@ -0,0 +1,184 @@
+    assertThat(datanodesInvolvedPerIteration).isLessThanOrEqualTo(number);
+    assertThat(metrics.getNumDatanodesInvolvedInLatestIteration()).isLessThanOrEqualTo(number);
+  }
+
+  @ParameterizedTest(name = "MockedSCM #{index}: {0}")
+  @MethodSource("createMockedSCMs")
+  public void balancerShouldObeyMaxSizeEnteringTargetLimit(@Nonnull MockedSCM mockedSCM) {
+    OzoneConfiguration ozoneConfig = mockedSCM.getOzoneConfig();
+    ozoneConfig.set("ozone.scm.container.size", "1MB");
+    ContainerBalancerConfiguration config = mockedSCM.getBalancerConfigByOzoneConfig(ozoneConfig);
+    config.setThreshold(10);
+    config.setMaxSizeToMovePerIteration(50 * STORAGE_UNIT);
+
+    // No containers should be selected when the limit is just 2 MB.
+    config.setMaxSizeEnteringTarget(2 * OzoneConsts.MB);
+    ContainerBalancerTask task = mockedSCM.startBalancerTask(config);
+
+    assertTrue(stillHaveUnbalancedNodes(task));
+    assertTrue(task.getContainerToSourceMap().isEmpty());
+
+    // Some containers should be selected when using default values.
+    ContainerBalancerConfiguration balancerConfig = mockedSCM.getBalancerConfigByOzoneConfig(new OzoneConfiguration());
+    balancerConfig.setBalancingInterval(1);
+
+    task = mockedSCM.startBalancerTask(balancerConfig);
+
+    // Balancer should have identified unbalanced nodes.
+    assertTrue(stillHaveUnbalancedNodes(task));
+    if (canNotBalanceAnyMore(task)) {
+      assertTrue(task.getContainerToSourceMap().isEmpty());
+    } else {
+      assertFalse(task.getContainerToSourceMap().isEmpty());
+    }
+  }
+
+  @ParameterizedTest(name = "MockedSCM #{index}: {0}")
+  @MethodSource("createMockedSCMs")
+  public void balancerShouldObeyMaxSizeLeavingSourceLimit(@Nonnull MockedSCM mockedSCM) {
+    OzoneConfiguration ozoneConfig = mockedSCM.getOzoneConfig();
+    ozoneConfig.set("ozone.scm.container.size", "1MB");
+    ContainerBalancerConfiguration config = mockedSCM.getBalancerConfigByOzoneConfig(ozoneConfig);
+    config.setThreshold(10);
+    config.setMaxSizeToMovePerIteration(50 * STORAGE_UNIT);
+
+    // No source containers should be selected when the limit is just 2 MB.
+    config.setMaxSizeLeavingSource(2 * OzoneConsts.MB);
+    ContainerBalancerTask task = mockedSCM.startBalancerTask(config);
+
+    assertTrue(stillHaveUnbalancedNodes(task));
+    assertTrue(task.getContainerToSourceMap().isEmpty());
+
+    // Some containers should be selected when using default values.
+    ContainerBalancerConfiguration newBalancerConfig =
+        mockedSCM.getBalancerConfigByOzoneConfig(new OzoneConfiguration());
+    newBalancerConfig.setBalancingInterval(1);
+
+    task = mockedSCM.startBalancerTask(newBalancerConfig);
+
+    // Balancer should have identified unbalanced nodes.
+    assertTrue(stillHaveUnbalancedNodes(task));
+
+    if (canNotBalanceAnyMore(task)) {
+      assertTrue(task.getContainerToSourceMap().isEmpty());
+      assertEquals(0, task.getSizeScheduledForMoveInLatestIteration());
```

Comment:
How about we add a short comment here explaining why size scheduled is expected to be 0 and map empty?